src/routes/safe/components/Apps/components/SignMessageModal/index.tsx

```diff
@@ -39,19 +42,36 @@ const convertToHumanReadableMessage = (message: string): string => {
   return humanReadableMessage
 }
 
-export const SignMessageModal = ({ message, isOpen, ...rest }: SignMessageModalProps): ReactElement => {
+export const SignMessageModal = ({ message, isOpen, method, ...rest }: SignMessageModalProps): ReactElement => {
   const web3 = getWeb3ReadOnly()
   const networkId = useSelector(currentChainId)
   const txRecipient = getSignMessageLibAddress(networkId) || ZERO_ADDRESS
-  const txData = getSignMessageLibContractInstance(web3, networkId)
-    .methods.signMessage(web3.eth.accounts.hashMessage(message))
-    .encodeABI()
-
-  const readableData = convertToHumanReadableMessage(message)
+  let txData, readableData
+  if (method == Methods.signMessage) {
+    txData = getSignMessageLibContractInstance(web3, networkId)
+      .methods.signMessage(web3.eth.accounts.hashMessage(message))
+      .encodeABI()
+    readableData = convertToHumanReadableMessage(message)
+  } else if (method == Methods.signTypedMessage) {
+    // Here we firstly convert the typed message to safe typed message which contains the safe address and the current chainId,
+    // then we hash it using ethers.utils._TypedDataEncoder.
+    txData = getSignMessageLibContractInstance(web3, networkId)
+      .methods.signMessage(
+        _TypedDataEncoder.hash(
+          { verifyingContract: rest.safeAddress, chainId: parseInt(networkId) },
+          { SafeMessage: [{ type: 'string', name: 'message' }] },
+          { message: message },
+        ),
+      )
+      .encodeABI()
+    readableData = JSON.stringify(JSON.parse(message), undefined, 4)
```

Comment:
Can it be that the JSON is invalid? I guess it would throw and crash the app then.

Reply:
Thanks for pointing it out! Now it will reject the tx and close the modal when the JSON is invalid or the method for visualizing is not supported.

Reply:
Looks good, thank you!